src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java

```diff
@@ -97,20 +117,29 @@ private JsonArray formatBatch(String topic) {
             }
 
             JsonElement recordJSON = recordToJSON(record);
-            JsonObject message = populateMetadata(topic, recordJSON, record.timestamp());
+            JsonObject message = populateMetadata(topic, recordJSON, record.timestamp(), () -> kafkaHeadersToJsonElement(record));
             batchRecords.add(message);
         }
 
         return batchRecords;
     }
 
+    private JsonElement kafkaHeadersToJsonElement(SinkRecord sinkRecord) {
+        Map<String, Object> headerMap = stream(sinkRecord.headers().spliterator(), false)
+                .collect(toMap(Header::key, Header::value));
+
+        String jsonString = gson.toJson(headerMap);
+
+        return gson.fromJson(jsonString, JsonElement.class);
```

Comment:
Could this map possibly contains a lot of keys and/or do we know about all of them before-hand or are they arbitrary?
I'm basically wondering if this `toJson/fromJson` couple could be turned in a more precise serialisation, directly creating a `JsonElement` instead.

Reply:
Headers are arbitrary and are created by the kafka producer. Some frameworks (for example Spring Kafka) have default headers that are added automatically to the kafka record. Headers are key-value pairs (string-bite[]) on the producer side, so a record could potentially have a lot of keys. We could add an additional property that would contain an explicit list of headers that the user would like to inject into DataDog.

I think that we can turn it into a more precise serialisation, will push it to the branch

Reply:
Hi @remeh, I have a question regarding the more precise serialization. I'm not so familiar with Gson, so I wanted to ask you if there is a more efficient way of doing the above conversion from `Header` to `JsonElement`. One approach that I found was to return a `JsonObject` and use toJsonTree method, but then we would serialize `N` times, where `N` is the number of `Header` objects. The above solution would have a constant (2) number of calls to the Gson library, no matter how many `Header` objects we have.

Reply:
> We could add an additional property that would contain an explicit list of headers that the user would like to inject into DataDog.

I'm not very aware of the general use-case of this PR but I think this addition could be interesting to have in the first iteration, yes. But then a new question enters the scene: what if the additional property isn't configured? Should we serialise all the headers? I think so if/since the user has explicitly set the `datadog.parse_record_headers` to `true`. Let me know if you have a different opinion.

>  One approach that I found was to return a JsonObject and use toJsonTree method, but then we would serialize N times, where N is the number of Header objects.

I don't think calling `N` times the serialisation is a problem, but it's probably not faster than calling it once for all headers anyway. Maybe you'll want to do the former (per header) when there is a filter and the latter (the whole object) when there is no filter?

Reply:
> I'm not very aware of the general use-case of this PR but I think this addition could be interesting to have in the first iteration, yes. But then a new question enters the scene: what if the additional property isn't configured? Should we serialise all the headers? I think so if/since the user has explicitly set the `datadog.parse_record_headers` to `true`. Let me know if you have a different opinion.
> 

This PR would solve the missing Kafka Record Headers in the DataDog Logs UI. For example, you publish two identical events into Kafka `{"foo": "foo", "bar": "bar"}`, that, downstream, produce multiple events that cannot be correlated in a deterministic way, and let's say you are not in a position to introduce an attribute in the event payload which would serve as a unique identifier. In that case, you can add a header containing a unique identifier (event payload would remain the same), and that header would be propagated to the downstream services and the downstream event would contain that header. The current DataDog Kafka Connector connector issue is that the headers are not parsed and sent to DataDog -> correlation between messages is almost impossible for the above scenario. 

Regarding the additional attribute, maybe we could go with the above approach for the first iteration. If the additional property is defined, then we would parse those headers that are explicitly stated, and if not we would parse them all.

Regarding the serialization, the current intention would be to have them as a `Map<String, Object>`, so one serialization call would be preferred from my perspective. The `toJson/fromJson` Gson methods were used because those were the only way I knew how to achieve the conversion from `Map<String, Object>` to JsonElement without serializing/deserializing multiple times.

Reply:
It makes sense to me, especially since this is behind a configuration flag and not enabled by default 👍.